Avoid warnings about redefined methods in EnforceSuperclass module

davidrunger opened the pull request:
Fix #1465

This change avoids the printing of warnings about some methods (`included`, `on_class`, and `on_send`) of the `RuboCop::Cop::EnforceSuperclass` module being redefined. These warnings are printed because, in addition to the definition of that module here in `rubocop-rails`, that module is first [also defined in `rubocop`][1]. (The module has been extracted to `rubocop-rails`, but is also left in `rubocop`, for backwards compatibility.)

[1]: https://github.com/rubocop/rubocop/blob/v1.73.2/lib/rubocop/cop/mixin/enforce_superclass.rb

This change avoids the warnings by first undefining the `RuboCop::Cop::EnforceSuperclass` module (the one from `rubocop`) before then redefining that module here in `rubocop-rails`.

-----------------

Before submitting the PR make sure the following are checked:

* [x] The PR relates to *only* one subject with a clear title and description in grammatically correct, complete sentences.
* [x] Wrote [good commit messages][1].
* [x] Commit message starts with `[Fix #issue-number]` (if the related issue exists).
* [x] Feature branch is up-to-date with `master` (if not - rebase it).
* [x] Squashed related commits together.
* [ ] Added tests. _(I am not sure we can easily test this?)_
* [x] Ran `bundle exec rake default`. It executes all tests and runs RuboCop on its own code.
* [x] Added an entry (file) to the [changelog folder](https://github.com/rubocop/rubocop-rails/blob/master/changelog/) named `{change_type}_{change_description}.md` if the new code introduces user-observable changes. See [changelog entry format](https://github.com/rubocop/rubocop/blob/master/CONTRIBUTING.md#changelog-entry-format) for details.
* [ ] ~~If this is a new cop, consider making a corresponding update to the [Rails Style Guide](https://github.com/rubocop/rails-style-guide).~~ _(Not applicable.)_

[1]: https://chris.beams.io/posts/git-commit/

davidrunger commented:
I'm not very sure that this is actually a good way to fix this issue, though it might be. But it feels sort of ugly/hacky to me. Maybe there is some more elegant way to fix the issue. I don't even really know specifically what caused this issue to emerge in version 2.30.0, which definitely limits my ability to conceive of a more elegant fix.

davidrunger commented:
I didn't add any tests, because I'm not really sure that we can easily test this? This repo [already has](https://github.com/rubocop/rubocop-rails/blob/v2.30.3/spec/spec_helper.rb#L27) RSpec configured with `config.warnings = true`, and yet no warnings are printed when running `bundle exec rspec`. I guess that the warnings appear only when `rubocop-rails` is required into another project?

davidrunger commented on lib/rubocop-rails.rb:
```diff
@@ -3,6 +3,7 @@
 require 'rack/utils'
 require 'active_support/inflector'
+String.remove_method(:blank?) if String.instance_method(:blank?)
 require 'active_support/core_ext/object/blank'
```
(This require statement is slightly misleading. It refers to `active_support/core_ext/object/blank`, but it not only patches `Object`, but also several other classes, [including `String`](https://github.com/rails/rails/blob/v8.0.1/activesupport/lib/active_support/core_ext/object/blank.rb#L135-L153).)

davidrunger commented on lib/rubocop/cop/mixin/enforce_superclass.rb:
```diff
@@ -1,7 +1,9 @@
 module RuboCop
-  module Cop
+  module Cop # rubocop:disable Style/Documentation
+    remove_const(:EnforceSuperclass) if defined?(EnforceSuperclass)
```
There are three warnings printed about methods of `EnforceSuperclass` being redefined. Rather than undefining each of those three methods individually, here I just undefine the whole `EnforceSuperclass` module (if it is already defined). We will then (re-)define it just below.

davidrunger commented on lib/rubocop/cop/mixin/enforce_superclass.rb:
```diff
@@ -1,7 +1,9 @@
 
 module RuboCop
-  module Cop
+  module Cop # rubocop:disable Style/Documentation
```
Because this `RuboCop::Cop` module is no longer purely just a wrapper/namespace, since below we are adding `remove_const(:EnforceSuperclass) if defined?(EnforceSuperclass)` to it, RuboCop now expects a documentation comment for it. But I don't think that we really want to document this module here, so I am disabling the cop here.

Earlopain commented:
How you handle `EnforceSuperclass` seems ok to me. RuboCop used to own this module but at some point it was moved to `rubocop-rails` and kept in rubocop itself for backwards compatibility.

Seems to be that that module here should have been scoped to a different namespace (but it wasn't, oh well). Can you add a comment that this module is also defined in RuboCop for backwards compatibility?

davidrunger commented:
> I openend https://github.com/rubocop/rubocop-rails/pull/1467 as an alternative for that.

@Earlopain Sounds good. I have removed from this PR the removal of String#blank?, leaving that to be fixed by #1467.

> Can you add a comment that this module is also defined in RuboCop for backwards compatibility?

Yes, good suggestion, thank you! I added this comment: https://github.com/rubocop/rubocop-rails/blob/f5c8ee8a33335fb08a8ccb9b1fe04bbaba2a2b25/lib/rubocop/cop/mixin/enforce_superclass.rb#L5-L8

koic commented:
Thanks!
